index.html

```diff
@@ -678,24 +679,31 @@ <h2>
                   to <a data-cite=
                   "payment-method-id#dfn-validate-a-payment-method-identifier">
                     validate a payment method identifier</a> with
-                    <var>paymentMethod</var>.<a data-lt=
-                    "PaymentMethodData.supportedMethods">supportedMethods</a>.
-                    If it returns false, then throw a <a>RangeError</a>
-                    exception. Optionally, inform the developer that the
-                    payment method identifier is invalid.
+                    <var>paymentMethod</var>.<a>supportedMethods</a>. If it
+                    returns false, then throw a <a>RangeError</a> exception.
+                    Optionally, inform the developer that the payment method
+                    identifier is invalid.
                   </li>
-                  <li>If the <a data-lt="PaymentMethodData.data">data</a>
-                  member of <var>paymentMethod</var> is missing, let
-                  <var>serializedData</var> be null. Otherwise, let
-                  <var>serializedData</var> be the result of
-                  <a>JSON-serializing</a>
-                    <var>paymentMethod</var>.<a data-lt="PaymentMethodData.data">data</a>
-                    into a string. Rethrow any exceptions.
+                  <li>If <var>paymentMethod</var>.<a>data</a> is present:
+                    <ol>
+                      <li>
+                        <a>JSON-serialize</a>
+                        <var>paymentMethod</var>.<a>data</a>. Rethrow any
+                        exceptions.
+                      </li>
+                      <li>If required by the specification that defines the
+                      <var>paymentMethod</var>.<a>supportedMethods</a>, replace
+                      <var>paymentMethod</var>.<a>data</a> with a
+                      <a>converted</a> IDL value of the type specified there
+                      (e.g., a <a data-cite=
+                      "payment-method-basic-card#dom-basiccardrequest">BasicCardRequest</a>
+                      in the case of [[?payment-method-basic-card]]). Rethrow
+                      any exceptions.
+                      </li>
+                    </ol>
                   </li>
-                  <li>Add the tuple (<var>paymentMethod</var>.<a data-lt=
-                  "PaymentMethodData.supportedMethods">supportedMethods</a>,
-                  <var>serializedData</var>) to
-                  <var>serializedMethodData</var>.
+                  <li>Append <var>paymentMethod</var> to
+                  <var>paymentMethods</var>.
```

Comment:
So the problem with not storing the JSON serialized version is that the JSON serialization now becomes meaningless. Consider the following:

```js
const bcr1 = { supportedNetworks: new Set("foo", "bar") };
```

With the new spec, the JSON-serialization will be `{ "supportedNetworks": {} }`. But the value stored in _paymentMethods_ will be `{ supportedNetworks: ["foo", "bar"] }` since Web IDL dictionary conversion converts all iterables to sequences.

This means that if someone, e.g., passed the JSON serialization across a process boundary and used it to attempt to reconstruct the result on the other side, they would not be able to match the spec. I thought our original spec was motivated exactly by that design.

Reply:
I believe the original design was motivated by some kind of security concern that was supposed to be mitigated by doing a JSON serialization on `.data` whose type is `object` to remove things like `Function` being added as properties.

Reply:
It may be good to check with implementations then. But a spec that JSON serializes then disregards the result seems incoherent.

Reply:
Gecko only does the IDL conversion, because we don't care about any payment method apart from "Basic Card". So, we don't store any `[[serializedMethodData]]` in our code... in fact, I don't see is doing any we JSON serialization of data at all in Gecko's code (@edenchuang?).

From the constructor, we check the method's details:
https://github.com/mozilla/gecko-dev/blob/master/dom/payments/PaymentRequest.cpp#L574  

We then just check `IsValidBasicCardRequest`: 
https://github.com/mozilla/gecko-dev/blob/master/dom/payments/PaymentRequest.cpp#L293

And then we convert it (to check throw if it fails):
https://github.com/mozilla/gecko-dev/blob/master/dom/payments/BasicCardPayment.cpp#L53

[WebKit does do the serialization](https://github.com/WebKit/webkit/blob/master/Source/WebCore/Modules/paymentrequest/PaymentRequest.cpp#L322) and stores it. However, @aestes also seems to suggest a [preference to not store the JSON serialization](https://github.com/w3c/payment-request/issues/813#issuecomment-441886747): "Browsers could convert from a JS object directly to a IDL dictionary rather than the current JS object > JSON string > JS object > IDL dictionary conversion algorithm".

Reply:
Thanks. @rsolomakhin?